### PR TITLE
fix: Only run postUpgradeTasks if package files are updated and flush changes to disk before executing

### PR DESCRIPTION
**Pull request by daekene**
This PR fixes two issues:
- Post-upgrade tasks are executed even if there are no updates to the package files. 
- Post-upgrade tasks are executed without any of the other changes Renovate is making are available on disk

Post-upgrade tasks are executed even if there are no updates to the package files:
This becomes a problem on the second Renovate run where no updates are necessary, but post upgrade tasks are still run. Any changes made by the post upgrade tasks are committed and the original changes to the package files that were in the branch are lost. This process repeats when Renovate is run again, seeing that the package file updates are lost, adds them again. The result is a constant stream of commits being added/removed, sometimes with or without the dependency updates. This has been fixed by only executing the tasks when there are other changes being made in the commit.

Post-upgrade tasks are executed without any of the other changes Renovate is making are available on disk:
It's currently impossible to write a post-upgrade task that incorporates the changes that Renovate is making on it's first run. The tasks only see the branch before any dependencies were updated, not after. This has been fixed by flushing out changed files to disk before executing the tasks.

### CODE REVIEW

File: lib/workers/branch/index.ts [hunk 1]
```diff
@@ -333,6 +340,22 @@ export async function processBranch(
...
+          config.updatedArtifacts
+        )) {
+          if (file.name !== '|delete|') {
+            let contents;
+            if (typeof file.contents === 'string') {
+              contents = Buffer.from(file.contents);
+            } else {
+              contents = file.contents;
+            }
+            await fs.outputFile(join(config.localDir, file.name), contents);
```
Comment (rarkins) [inline lib/workers/branch/index.ts:355]: I think this shouldn't be necessary for updated artifacts as they are always updated on disk by third party tools anyway. However, probably little harm either.

Comment (rarkins) [reply lib/workers/branch/index.ts:355]: Please use our util/fs library instead: https://github.com/renovatebot/renovate/blob/f97b19451539e5940258defdbc0afd4754df0be7/lib/util/fs.ts#L43

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/branch/index.ts [hunk 2]
```diff
@@ -319,6 +322,10 @@ export async function processBranch(
     }
 
     if (
+      /* Only run post-upgrade tasks if there are changes to package files... */
+      ((config.updatedPackageFiles && config.updatedPackageFiles.length > 0) ||
```
Comment (viceice) [inline lib/workers/branch/index.ts:326]:
```suggestion
      ((config.updatedPackageFiles?.length > 0) ||
```

File: lib/workers/branch/index.ts [hunk 3]
```diff
@@ -319,6 +322,10 @@ export async function processBranch(
...
+        /* ... or changes to artifacts */
+        (config.updatedArtifacts && config.updatedArtifacts.length > 0)) &&
```
Comment (viceice) [inline lib/workers/branch/index.ts:328]:
```suggestion
        (config.updatedArtifacts?.length > 0)) &&
```

**Review by viceice — CHANGES_REQUESTED** [review]
Some small improvments

File: lib/workers/branch/index.ts (hunk 1 above)
Comment (daekene) [reply lib/workers/branch/index.ts]: I had the same thought about updated artifacts and I agree that there's probably little harm to include it. I've replaced `fs.outputFile` with `writeLocalFile`

**Pull request merged**